examples/platform_channel/linux/my_application.cc

```diff
@@ -0,0 +1,262 @@
+#include <math.h>
+#include <upower.h>
+
+#include "my_application.h"
+
+#include <flutter_linux/flutter_linux.h>
+#ifdef GDK_WINDOWING_X11
+#include <gdk/gdkx.h>
+#endif
+
+#include "flutter/generated_plugin_registrant.h"
+
+struct _MyApplication {
+  GtkApplication parent_instance;
+  char** dart_entrypoint_arguments;
+
+  FlView* view;
+
+  // Connection to UPower.
+  UpClient* up_client;
+  GPtrArray* battery_devices;
+
+  // Channel for Dart code to request the battery information.
+  FlMethodChannel *battery_channel;
+
+  // Channel to send updates to Dart code about battery charging state.
+  FlEventChannel *charging_channel;
+  gchar* last_charge_event;
+  bool emit_charge_events;
+};
+
+G_DEFINE_TYPE(MyApplication, my_application, GTK_TYPE_APPLICATION)
+
+// Check the charging state and emit an event if necessary.
+static void update_charging_state(MyApplication *self) {
+  bool charging = false;
+  for (guint i = 0; i < self->battery_devices->len; i++) {
+    UpDevice* device = static_cast<UpDevice*>(g_ptr_array_index(self->battery_devices, i));
+
+    guint state;
+    g_object_get(device, "state", &state, nullptr);
+    if (state == UP_DEVICE_STATE_CHARGING || state == UP_DEVICE_STATE_FULLY_CHARGED) {
+      charging = true;
+    }
+  }
+
+  const gchar* charge_event = charging ? "charging" : "discharging";
```

Comment:
Optional nit: would it make sense to eliminate the bool and just set this directly? Each state is only set in one place so there wouldn't be any string duplication.